### PR TITLE
fix(cosh-ng): [shell] pair dropped extension audit

### LINKED ISSUE #2093: [TestAgent][cosh-ng] bug: note_capped_run 在审批未展示前写 approval.requested,被 turn_extension 静默丢弃留孤儿审计
### Component

cosh-ng

### Bug Description

`note_capped_run` 在 agent 跑满 turn 上限时,先调 `audit.record_approval_requested` 写 `approval.requested` 审计事件,然后把 approval 塞进 `pending_turn_extension`。但 `activate_pending_turn_extension` 在多个早退分支(还有 active run / 队列非空 / 有 pending interaction / compaction pending)里直接 `pending_turn_extension = None`(line 75)或返回 false(line 79)丢弃 pending,从不在这些分支调 `record_approval_resolved`,留孤儿 `approval.requested` 没有 `approval.resolved` 配对,审计链路不闭合。

### Steps to Reproduce

```shell
ssh root@<ECS-IP>
# 静态确认:看 note_capped_run 写 approval.requested
sed -n '23,60p' src/cosh-ng/crates/cosh-shell/src/agent/turn_extension.rs
# 静态确认:看 activate 的早退分支丢弃 pending 不写 approval.resolved
sed -n '64,91p' src/cosh-ng/crates/cosh-shell/src/agent/turn_extension.rs
```

> 注:本 bug 为静态源码确认,动态复现需 turn cap + LLM 触发 `note_capped_run` 后命中早退分支。

### Actual Behavior

`src/cosh-ng/crates/cosh-shell/src/agent/turn_extension.rs:51-58` 在 `note_capped_run` 里写 `approval.requested`:

```rust
if let Some(audit) = state.audit.as_mut() {
    approval.audit_ref = audit.record_approval_requested(approval_audit_input(&approval));
}
state.agent_run.pending_turn_extension = Some(PendingTurnExtension {
    request: continuation_request(&active_run.request, &approval_id),
    approval,
    provider_session_id,
});
```

`activate_pending_turn_extension`(line 64-91)在多个早退分支直接丢弃 pending 不写 `approval.resolved`:

```rust
if state.agent_run.active.is_some() || !state.agent_run.queued_requests.is_empty() {
    state.agent_run.pending_turn_extension = None;  // 丢弃,无 approval.resolved
    return Ok(false);
}
if state_has_pending_interaction(state) {
    return Ok(false);  // 丢弃路径,无 approval.resolved
}
```

孤儿 `approval.requested` 留在审计日志,无配对的 `approval.resolved`。

### Expected Behavior

丢弃 pending 前应调 `audit.record_approval_resolved(...status="cancelled"...)` 写 `approval.resolved` 配对,审计链路闭合;或在 `note_capped_run` 写 `approval.requested` 时延后到 `activate` 真正渲染后再写。

### Root Cause

`turn_extension.rs:51-58` 写 `approval.requested` 早于审批实际渲染;`activate_pending_turn_extension` 在 line 75 / 79 等早退分支直接丢 pending,未调 `record_approval_resolved`,留孤儿。触发条件:agent 跑满 turn 上限 → `note_capped_run` 写 `approval.requested` → 进入 `activate_pending_turn_extension` 命中早退分支(还有 active run / 队列非空 / pending interaction / compaction pending)。设计取舍:无文档说明为何丢弃时不补 resolved;非 by-design,是审计事件配对疏漏。

### Suggested Fix

在 `activate_pending_turn_extension` 的丢弃分支(line 75、line 79)里,丢 pending 前调 `state.audit.record_approval_resolved(approval_audit_input(&pending.approval) with status="cancelled")`;或在 `note_capped_run` 不立刻写 `approval.requested`,延后到 `activate` 真正渲染审批时再写,从源头避免孤儿。

### Environment

- OS: Alibaba Alinux 4, kernel 6.6.102
- Component: cosh-ng
- Package/RPM: cosh-ng-0.13.0-2.alnx4

### Additional Context

- 相关 issue/PR: 无
- 稳定性: 静态源码确认(动态复现需 turn cap + LLM)
- 被测版本 main HEAD: a96622bd

### PR DESCRIPTION
**Pull request by xuchunmei000**
## Why

When an agent run exhausts its turn budget, `note_capped_run` records an
`approval.requested` audit event before the approval panel is rendered and
parks the approval in `pending_turn_extension`. If newer user activity
(an active run or queued requests) supersedes the pending extension,
`activate_pending_turn_extension` silently discarded it without recording
a matching `approval.resolved`, leaving an orphan `approval.requested` in
the audit log and breaking request/resolution pairing.

## What changed

`activate_pending_turn_extension` now withdraws a superseded pending
extension through a new `discard_pending_turn_extension` helper: the
approval status is set to `Cancelled` and a paired `approval.resolved`
(outcome `cancelled`) is recorded before the pending entry is dropped.
A durability failure only degrades the recorder — nothing executes on
this path, so there is no boundary to fail closed on. The alternative
(deferring `approval.requested` to render time) was rejected because it
would shift the `audit_ref` / owned-approvals ownership contract in
`note_capped_run`.

The compaction and pending-interaction early-exit branches are unchanged:
they retain the pending extension for later activation and do not drop it.

## Related issue

closes #2093

## User / Agent impact

None in interactive behavior. Audit consumers now see a paired
`approval.resolved` (status `cancelled`) instead of an orphan
`approval.requested` when a pending turn extension is superseded.

## Risk and compatibility

- [ ] Public CLI, API, configuration, or documented behavior changed
- [x] Privileged or security-sensitive behavior changed
- [ ] Cross-component contract changed
- [ ] Migration or rollback guidance is needed

The audit trail gains one additional `approval.resolved` event on an
already-existing drop path; the event uses the established
`"cancelled"` mapping in `ShellAuditRecorder::record_approval_resolved`,
so no schema or consumer changes are required.

## Validation

Environment: Linux (Alinux), `src/cosh-ng` workspace.

- `cargo fmt --all -- --check` — clean
- `cargo clippy -p cosh-shell --all-targets -- -D warnings` — clean
- `cargo test -p cosh-shell turn_extension` — 8 passed (6 unit tests
  including the new
  `superseded_pending_extension_closes_audit_with_cancelled_resolution`,
  which asserts the persisted segment contains the paired
  `approval.resolved` with `outcome.status = cancelled`, plus 2 raw_cli
  end-to-end cases)

Edge cases covered: superseding via a new active run (drop branch);
existing tests keep covering the no-session and non-capped-failure
paths. 

## Documentation and rollback

No documentation changes (internal audit pairing fix). Rollback: revert
the single commit; no data migration involved.

### CODE REVIEW

File: src/cosh-ng/crates/cosh-shell/src/agent/turn_extension.rs [hunk 1]
```diff
@@ -271,6 +286,66 @@ mod tests {
...
+        std::fs::create_dir(&root).unwrap();
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::PermissionsExt;
+            std::fs::set_permissions(&root, std::fs::Permissions::from_mode(0o700)).unwrap();
+        }
+        root.canonicalize().unwrap()
+    }
+
+    fn segment_text(root: &std::path::Path) -> String {
```
[the remark this answers is not in the document]
Comment (xuchunmei000) [reply src/cosh-ng/crates/cosh-shell/src/agent/turn_extension.rs:346]: Fixed in the amended commit. segment_text now uses expect() with messages stating the expected v1/segments/<date> layout, so a future on-disk layout change fails with an actionable signal instead of a bare unwrap panic.

File: src/cosh-ng/crates/cosh-shell/src/agent/turn_extension.rs [hunk 2]
```diff
@@ -90,6 +90,21 @@ pub(crate) fn activate_pending_turn_extension<W: Write>(
     Ok(true)
 }
 
+/// Withdraws a recorded extension that newer user activity superseded before
```
[the remark this answers is not in the document]
Comment (xuchunmei000) [reply src/cosh-ng/crates/cosh-shell/src/agent/turn_extension.rs:93]: Fixed in the amended commit. The parked copy is currently guaranteed Pending by its single construction site, so this cannot fire today, but the guard makes the helper self-contained — discard now early-returns without recording when the status is not Pending, with a comment documenting the invariant.

**Review by kongche-jbw — COMMENTED** [review]
Review baseline: `8f2416470233...b76e6ce5cc49`

未发现阻塞问题。

Remaining risks:
- 未做真实 LLM turn-cap 竞态黑盒复现；审计写入故障时仍可能无法形成完整事件对。

Verification scope:
- 审查完整 diff、相邻审批/审计/运行生命周期代码及全部现有 review threads。
- 本地 `cargo test -p cosh-shell turn_extension` 通过：6 个 unit、2 个 raw_cli。
- 本地格式检查通过；当前 head 的 cosh-ng CI、fast checks 和 release build 均通过。

**Review by kongche-jbw — APPROVED** [review]
